### PR TITLE
Harden Boost data handoffs and graph cache writes

### LINKED ISSUE #7853: Harden Boost data handoffs and graph cache writes
Boost handed values to the graph cache and to rrdtool without validating them at the boundary. They are now checked where they enter and fail closed when unusable.
Includes a schema change (`cacti.sql`, `docs/audit_schema.sql`) for the audit columns, so this one needs an upgrade path review.
## Release targeting
- Target branch: `develop`
- Planned milestone: `v1.3.0`
- Release line: primary development branch; this is not a 1.2.x backport.

---

Tracking issue for #7729, which carries the fix.

### PR DESCRIPTION
**Pull request by somethingwithproof**
Boost handed values to the graph cache and to rrdtool without validating them at the boundary. They are now checked where they enter and fail closed when unusable.

Includes a schema change (`cacti.sql`, `docs/audit_schema.sql`) for the audit columns, with a guarded `1_3_0` upgrade path.

## Release targeting

- Target branch: `develop`
- Planned milestone: `v1.3.0`
- Release line: primary development branch; this is not a 1.2.x backport.

Closes #7853

### Review and rebase status

Rebased onto current `develop` at `770c503f3`. The upgrade conflict was resolved by preserving the current process-lock table creation while retaining the corrected uppercase `NULL` schema keys.

All existing Copilot and human review findings are addressed and every review conversation is resolved. The final changes also remove the unreachable negative aggregate branch, recreate the Boost completion table when an upgrade finds it missing, preserve in-flight completion rows during upgrade, execute current production SQL in the MariaDB integration proof, and guard nullable graph output before atomic cache publication.

### Verification

- Docker focused suite: 43 tests, 215 assertions
- Docker real-MariaDB integration: 7 tests, 27 assertions
- Expanded Boost suite: 112 fast Docker tests / 469 assertions plus all 15 real-MariaDB 11.4 cases / 43 assertions
- Added coverage for packet-limit chunking, non-destructive Windows cache replacement and cleanup failure paths, missing-table reconstruction, runtime legacy-schema repair, upgrade idempotence, and duplicate child-completion enforcement
- Complete JavaScript suite: 15 tests passed; PHP-CS-Fixer, PHP syntax, sink inventory, and architectural hotspot guards pass
- Strict-SQL portability: upgrade integration cases pass with MariaDB `ANSI_QUOTES` enabled
- Fault-injected chunk failure now stops subsequent staging; focused packet/cache suite passes 15 tests / 45 assertions
- Recovery status SQL is covered for MariaDB `ANSI_QUOTES` compatibility
- All recovery/spike-kill SQL literals are verified under MariaDB `ANSI_QUOTES`
- Cache publication preserves the established `0644` reader contract across separate poller/web users
- Cache publication fails closed if the required reader mode cannot be applied
- Parent and child startup both validate/repair the completion schema before truncation or completion writes
- Recovery fetches remain strictly bounded even when more than 150,000 rows share the cutoff timestamp
- Full Docker PHPStan passes across all 436 analyzed files
- Failed recovery restores the poller from status 5 while preserving failure logs and exit status
- Rejected archive-like table warnings include a control-character-sanitized table name
- Unit tests rely on the suite bootstrap instead of redeclaring core helpers
- Generated graph-cache secrets are re-read directly from the database after `INSERT IGNORE`, avoiding stale option-cache values during concurrent creation
- Cache directories trim either Windows or Unix trailing separators before appending the platform-native separator
- Hand-off test helpers are guarded against redeclaration during repeated Pest collection
- MariaDB confirms `child_id` is the second column of the composite `run_child` unique key; the audit schema correctly leaves its standalone `Key` field blank
- Full GitHub CI is green on head `fbbf8c50c`; the latest Copilot review covers all 21 changed files with 0 new comments

### CODE REVIEW

**Comment by somethingwithproof** [comment]
The coverage gate isn't a coverage problem. Pest exits cleanly part-way through the run, so `coverage.xml` is never written and the gate reports it as missing.

Reproduced locally with the CI command:

- on `develop`: 1616 passed, writes a 2.2 MB `coverage.xml`
- on this branch: output stops after `integration/Issue7343InstallerCsrfRecoveryIntegrationTest`, exit code 0, no `coverage.xml`, no summary line and no error

The next file that should run is `integration/Issue7427UpdateHostStatusByIdIntegrationTest`, which passes on its own. It exits rather than crashing, which is why nothing shows up as a failed test.

One thing that looks related: `tests/integration/BoostFlushOutputBatchTest.php` is the first integration test to pull in `tests/Helpers/UnitStubs.php`. Those stubs guard-define functions and constants (`cacti_sizeof`, `read_user_setting`, `POLLER_VERBOSITY_DEBUG`) that `lib/functions.php` and `include/global_constants.php` declare unguarded, so whichever loads second aborts the process. Loading the two files together in either order reproduces a redeclare fatal.

File: poller_boost.php [hunk 1]
```diff
@@ -755,7 +793,9 @@ function boost_output_rrd_data(int $child) : mixed {
 	}
 
 	if ($total_rows == 0) {
-		return 0;
+		rrd_close($rrdtool_pipe);
+
+		return -1;
 	}
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller_boost.php:810]: Acknowledged. This concern remains open and needs a follow-up change; I am leaving the thread unresolved until it is addressed.

File: poller.php [hunk 2]
```diff
@@ -1132,8 +1132,11 @@ function poller_table_maintenance() : void {
 	if (!db_table_exists('poller_output_boost_processes')) {
 		db_execute('CREATE TABLE  `poller_output_boost_processes` (
 			`sock_int_value` bigint(20) unsigned NOT NULL auto_increment,
+			`run_id` char(32) NOT NULL default "",
+			`child_id` int(10) unsigned NOT NULL default "0",
 			`status` varchar(255) default NULL,
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller.php]: Acknowledged. This concern remains open and needs a follow-up change; I am leaving the thread unresolved until it is addressed.

File: cacti.sql [hunk 3]
```diff
@@ -2572,8 +2572,11 @@ CREATE TABLE `poller_output_boost_local_data_ids` (
 
 CREATE TABLE `poller_output_boost_processes` (
   `sock_int_value` bigint(20) unsigned NOT NULL auto_increment,
+  `run_id` char(32) NOT NULL default '',
+  `child_id` int(10) unsigned NOT NULL default '0',
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply cacti.sql]: Acknowledged. This concern remains open and needs a follow-up change; I am leaving the thread unresolved until it is addressed.

File: poller_recovery.php [hunk 4]
```diff
@@ -224,64 +241,57 @@ function debug(string $string) : void {
...
-					if (($packet_size + $sql_size) >= $max_allowed_packet) {
-						$record_count = cacti_sizeof($sql_array);
+				$sql_array = [];
 
-						cacti_log('RECOVERY: Writing ' . $record_count . ' records (' . $packet_size . ' bytes) to main (partial).', false, 'POLLER');
+				foreach ($rows as $r) {
+					$sql_array[] = '(' . (int) $r['local_data_id'] . ',' .
+						db_qstr($r['rrd_name'], $remote_db_cnn_id) . ',' .
+						db_qstr($r['time'], $remote_db_cnn_id) . ',' .
+						db_qstr($r['output'], $remote_db_cnn_id) . ')';
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller_recovery.php]: Acknowledged. This concern remains open and needs a follow-up change; I am leaving the thread unresolved until it is addressed.

File: lib/boost.php [hunk 5]
```diff
@@ -306,14 +369,17 @@ function boost_poller_on_demand(array &$results) : bool {
...
 
 		if (boost_check_correct_enabled()) {
-			// if boost redirect is on, rows are being inserted directly
-			if (read_config_option('boost_redirect') == 'on') {
-				restore_error_handler();
-
-				return false;
-			}
-
 			if (cacti_sizeof($results)) {
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply lib/boost.php:671]: Acknowledged. This concern remains open and needs a follow-up change; I am leaving the thread unresolved until it is addressed.

File: poller_boost.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller_boost.php:810]: Fixed in e2fbb8be1. An empty Boost shard now returns zero, so the parent treats it as successful no-op work rather than a failed child.

File: poller.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller.php]: Fixed in e2fbb8be1. The fallback DDL now uses a single-quoted empty string default for run_id and an unquoted numeric default for child_id, avoiding ANSI_QUOTES sensitivity.

File: lib/boost.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply lib/boost.php:671]: Fixed in e2fbb8be1. boost_poller_on_demand now exits before staging when boost_redirect is enabled and restores the caller error handler and reporting level first.

File: poller_recovery.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller_recovery.php]: Fixed in e2fbb8be1. Recovery transfers and deletes bounded 1000-row chunks instead of materializing a second tuple array for the full 150000-row fetch, while retaining rows on any failed acknowledgement.

File: cacti.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply cacti.sql]: Fixed in e2fbb8be1. The schema now defines child_id with the unquoted numeric default 0.

File: install/upgrades/1_3_0.php [hunk 6]
```diff
@@ -90,6 +90,10 @@ function upgrade_to_1_3_0() : void {
 	db_install_add_key('host', 'INDEX', 'host_template_id', ['host_template_id']);
 	db_install_add_key('sessions', 'INDEX', 'user_id', ['user_id']);
 	db_install_add_key('sessions', 'INDEX', 'access', ['access']);
+	db_install_execute('TRUNCATE TABLE poller_output_boost_processes');
```
Comment (bmfmancini) [inline install/upgrades/1_3_0.php]: Why are we truncating boost processes here ?

If the user is updating and a boost process is running I am pretty sure this would cause boost to fail

File: lib/boost.php [hunk 7]
```diff
@@ -245,24 +243,88 @@ function boost_flush_output_batch(array $value_tuples, mixed $conn = false) : vo
...
+			$success = false;
+		} elseif (db_affected_rows($conn) < $rows) {
+			cacti_log('WARNING: Boost staging ignored one or more duplicate sample keys.', false, 'BOOST');
+		}
+	}
+
+	return $success;
+}
+
+/** Confirm that every handed-off data source belongs to the claiming poller. */
```
Comment (bmfmancini) [inline lib/boost.php]: Is this check required ?
Boost as its current setup only runs on the main poller
So I dont see the need to check the Poller id of the data source

File: install/upgrades/1_3_0.php (hunk 6 above)
Comment (somethingwithproof) [reply install/upgrades/1_3_0.php]: You are right, nothing stops cron starting poller_boost.php while the installer runs, and the truncate would take out the completion rows the parent counts in its drain loop. It was only there so the run_child UNIQUE key could build over pre-1.3.0 rows, which the ALTERs collapse onto ('', 0). Changed to delete just those, so a child of an in-flight run keeps its record. Integration test covers both the 1.2.x table and the mid-run case.

File: lib/boost.php (hunk 7 above)
Comment (somethingwithproof) [reply lib/boost.php]: Correct, and it was worse than unused. poller.php forces boost_redirect on whenever a second collector exists, so boost_poller_on_demand() returns at the redirect check before that line is reached. The matching check in poller_recovery.php was reachable and turned an ordinary device deletion on main into a collector stuck in recovery, since the exact-set match fails forever on rows whose data source is gone. Both removed.

**Comment by somethingwithproof** [comment]
Rebased onto current `develop` (`770c503f3`) and resolved the `install/upgrades/1_3_0.php` conflict by preserving the new process-lock table while retaining the uppercase `NULL` schema keys. Docker validation after the rebase: 43 focused tests / 215 assertions, plus 7 real-MariaDB Boost integration tests / 27 assertions. All existing review conversations remain resolved. CI has been retriggered on head `45eca7618`.

File: tests/Unit/Database/UpgradeColumnNullKeyTest.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+test('upgrade column definitions spell the NULL key in upper case', function () {
+	$upgrades = glob(dirname(__DIR__, 3) . '/install/upgrades/*.php');
+
+	expect($upgrades)->not->toBeEmpty();
+
+	foreach ($upgrades as $upgrade) {
+		// toContain() takes needles, not a message.
+		expect(str_contains(file_get_contents($upgrade), "'null' =>"))
+			->toBeFalse();
+	}
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply tests/Unit/Database/UpgradeColumnNullKeyTest.php:27]: Fixed in 66d26029b. Both file reads now assert success before their contents are inspected, so a missing or unreadable source fails at the read boundary with a clear expectation.

File: lib/boost.php [hunk 9]
```diff
@@ -469,6 +499,87 @@ function boost_return_cached_image(&$graph_data_array) : bool {
...
+
+		if ($result === false || $result === 0) {
+			fclose($fileptr);
+			@unlink($temp_file);
+
+			return false;
+		}
+
+		$written += $result;
+	}
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply lib/boost.php:931]: Fixed in 66d26029b. Atomic cache writes now slice the output into bounded 1 MiB chunks and handle partial writes within each bounded chunk. The Docker regression writes and verifies a payload spanning more than three chunks.

File: tests/Unit/Database/UpgradeColumnNullKeyTest.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+	expect($upgrades)->not->toBeEmpty();
+
+	foreach ($upgrades as $upgrade) {
+		$contents = file_get_contents($upgrade);
+
+		expect($contents)->not->toBeFalse();
+
+		// toContain() takes needles, not a message.
+		expect(str_contains($contents, "'null' =>"))
+			->toBeFalse();
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply tests/Unit/Database/UpgradeColumnNullKeyTest.php]: Fixed in d44d22534. The guard now uses a regex that catches either quote style and arbitrary spacing before the array arrow. The pre-fix Docker reproduction showed the compact double-quoted form was missed; the updated focused Docker suite passes (19 tests, 161 assertions).

**Comment by somethingwithproof** [comment]
Also addressed the two suppressed findings from the latest Copilot summary in d44d22534: the Boost source helper now throws a clear RuntimeException when a source file cannot be read, and Windows cache refreshes fall back to replacing an existing destination after rename-over-existing fails. Docker reproduction was performed before the edits (including simulated Windows rename semantics), and the resulting focused Docker suite passes 19 tests / 161 assertions. The exact files also pass PHP-CS-Fixer and PHP 8.3 syntax checks.

**Comment by somethingwithproof** [comment]
Follow-up on the two suppressed observations in the d44d22534 Copilot review: I reproduced both against MariaDB 11.4 in a disposable Docker container. No code change is appropriate. For recovery deletion, EXPLAIN reports the same PRIMARY-key range plan (key length 91, two rows, Using where) for both the current bounded OR chain and the suggested row-constructor IN form. For audit_schema.sql, information_schema.STATISTICS for the actual poller_output_boost_processes MEMORY schema reports run_child as NON_UNIQUE=0 for both columns and INDEX_TYPE=HASH, exactly matching the checked-in audit rows; changing the second field to a non-zero value would incorrectly describe a UNIQUE index as non-unique. The container was removed after verification.

**Comment by somethingwithproof** [comment]
Final readiness audit: the complete GitHub matrix is green on d44d22534, Docker regressions pass, every review conversation is resolved, the branch contains the latest develop base, and GitHub reports it mergeable with no conflicts. A standard exact-head squash merge was attempted, but branch policy rejected it because required human approval is outstanding; no administrator bypass was used.

**Comment by somethingwithproof** [comment]
Copilot strict-SQL feedback is fixed in 4062bb019. The original DDL failure was reproduced first in MariaDB 11.4 Docker with ANSI_QUOTES (ERROR 1054), then the fixture was corrected to use SQL string literals and the integration setup now explicitly enables ANSI_QUOTES. All five process-table upgrade cases pass under that mode; formatting and syntax checks also pass. The push made the inline thread outdated, leaving no unresolved conversation.

File: lib/boost.php [hunk 11]
```diff
@@ -245,24 +243,50 @@ function boost_flush_output_batch(array $value_tuples, mixed $conn = false) : vo
...
+				$acknowledged = db_execute($sql_prefix . $out_buffer, true, $conn) !== false;
+			} catch (Throwable) {
+				$acknowledged = false;
+			}
+
+			if (!$acknowledged) {
+				$success = false;
+			} elseif (db_affected_rows($conn) < $rows) {
+				cacti_log('WARNING: Boost staging ignored one or more duplicate sample keys.', false, 'BOOST');
+			}
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply lib/boost.php:586]: Fixed in ea33a64d5. A fault-injecting PDO reproduced the issue in Docker: the old code attempted a third INSERT after the second chunk failed. boost_flush_output_batch() now returns immediately on the first unacknowledged chunk. The regression proves only the first acknowledged row remains and no later INSERT is attempted; all 106 Boost-related fast tests / 445 assertions pass.

**Comment by somethingwithproof** [comment]
The final suppressed Copilot finding is fixed in 365f6f169. The original poller status UPDATE was reproduced under MariaDB 11.4 ANSI_QUOTES and failed with ERROR 1054 because "2" was interpreted as an identifier. Recovery now writes the integer as status=2, with a handoff regression. Docker tests, PHP-CS-Fixer, syntax, sink inventory, and architectural hotspot checks pass. This review generated no inline thread, so there was no conversation to resolve.

File: tests/HandOff/BoostDataHandOffTest.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	expect($recovery)->toContain('array_slice($rows, $offset, $transfer_chunk_size)');
+	expect($recovery)->toContain('recovery_transfer_rows($chunk, $remote_db_cnn_id, $local_db_cnn_id)');
+	expect($recovery)->not->toContain('DELETE FROM poller_output_boost WHERE time <=');
+});
+
+test('Recovery status SQL remains valid when ANSI_QUOTES is enabled', function () {
+	$recovery = boostSource('poller_recovery.php');
+
+	expect($recovery)->toContain('SET status=2')
+		->and($recovery)->not->toContain('SET status="2"');
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply tests/HandOff/BoostDataHandOffTest.php]: Fixed in 47314b7fa. Before editing, MariaDB 11.4 Docker reproduced failures for recovery_pid, status=5, spike-kill U, and the disabled-site predicate under ANSI_QUOTES. All now use numeric literals, PHP SQL string literals, or db_qstr() as appropriate. The corrected forms were re-executed successfully under ANSI_QUOTES, and the handoff guard now covers all of them.

**Comment by somethingwithproof** [comment]
Fresh review found one final fail-closed cache-publication edge case. I reproduced it before editing in the Docker PHP 8.4 test environment: when the replacement source could not be renamed, the old cache file was deleted and the new regression failed (12 passed, 1 failed). Commit `a308a6461` now moves the old object to a same-directory backup, publishes the replacement, and restores the old object if publication fails. Post-fix Docker results: the focused file passes 13 tests / 36 assertions, the complete fast Boost selection passes 106 tests / 357 assertions, PHP syntax and PHP-CS-Fixer pass, and both security guard suites pass. The full hosted matrix and a fresh automated review are running on the exact head.

**Comment by somethingwithproof** [comment]
The cache-permission compatibility finding is fixed in commit `b1eb88ade`. Before editing production code, Docker reproduced the regression with an explicit mode assertion: the cache was created as `0640` (416 decimal) instead of the established `0644` (420 decimal), so a web/PHP user outside the poller group could lose read access. The atomic writer and both executable/contract tests now require `0644`, matching `lib/rrd.php` and `poller_realtime.php`. Post-fix Docker results are 13 focused tests / 37 assertions and 106 fast Boost tests / 358 assertions; PHP-CS-Fixer, PHP syntax, sink inventory, and architectural hotspot guards also pass. Full CI and a fresh automated re-review are running on the exact head.

**Comment by somethingwithproof** [comment]
The two latest operational findings are fixed in commit `9d909248c`.

Before changing production code, Docker MariaDB 11.4 reproduced the child completion failure against the legacy process table: the insert failed with `ERROR 1054 (42S22): Unknown column 'run_id'`. Runtime startup now creates or upgrades the completion table, removes only identifier-less legacy rows before adding the unique key, preserves valid in-flight rows, rechecks concurrent installer repairs, and fails closed before either parent truncation or child work if the schema is still unusable. The new real-database case executes the repair and then the exact child upsert successfully.

The cache writer also now checks both `fflush()` and `chmod(0644)` before publication, cleans the temporary file, and returns failure when either contract cannot be established.

Post-fix Docker results: 107 fast Boost tests / 364 assertions and all 15 MariaDB-backed cases / 43 assertions. PHP-CS-Fixer, PHP syntax, sink inventory, architectural hotspot guards, and staged-source guard pass. The disposable database container was removed. Full CI and a fresh automated re-review are running on the exact head.

File: lib/boost.php [hunk 13]
```diff
@@ -469,6 +502,127 @@ function boost_return_cached_image(&$graph_data_array) : bool {
...
+
+	$flushed = fflush($fileptr);
+	fclose($fileptr);
+	chmod($temp_file, 0644);
+
+	$published = $flushed && @rename($temp_file, $cache_file);
+
+	if (!$published && $flushed && PHP_OS_FAMILY === 'Windows') {
+		$published = boost_replace_cache_file_on_windows($temp_file, $cache_file);
+	}
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply lib/boost.php:946]: Fixed in `9d909248c`. Publication now checks both `fflush()` and `chmod(0644)` and removes the temporary object without renaming it when either operation fails. The executable cache test and handoff contract cover the required mode; Docker passes the focused and broad Boost suites.

**Comment by somethingwithproof** [comment]
The recovery batching finding is fixed in commit `b15e924c3`. Before editing, MariaDB 11.4 Docker loaded 150,005 rows with the same timestamp and reproduced the current query fetching all 150,005 despite a configured limit of 150,000. The production fetch now applies an integer-formatted SQL `LIMIT` after its deterministic timestamp/data-ID ordering; the same Docker dataset returns exactly 150,000 and leaves five rows for the next loop. The handoff contract now enforces the bounded query. Post-fix results: 11 focused recovery/handoff tests / 53 assertions and 107 fast Boost tests / 365 assertions; formatter, syntax, sink inventory, architectural hotspot, and staged-source guards pass. The disposable database was removed, and full CI plus another automated re-review are running on the exact head.

File: poller_recovery.php [hunk 14]
```diff
@@ -224,64 +275,37 @@ function debug(string $string) : void {
 				FROM poller_output_boost
 				WHERE time <= ?
 				ORDER BY time ASC, local_data_id ASC',
-				[$max_time]);
+				[$max_time], true, $local_db_cnn_id);
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply poller_recovery.php]: Fixed in `b15e924c3`. MariaDB Docker reproduced 150,005 rows being fetched from 150,005 tied timestamps with a 150,000 limit. The ordered production fetch now includes an integer-formatted `LIMIT`; the same dataset returns exactly 150,000, and the handoff contract plus full fast Boost suite pass.

**Comment by somethingwithproof** [comment]
The failed hosted PHPStan job was reproduced locally in the Docker test image before editing (four purity-inference errors plus the new `getmypid()` type error). Commit `f8f979e14` keeps schema failures fail-closed but removes redundant post-failure probes and casts the PID at the new unregister call. Docker PHPStan now analyzes all 436 files with no errors. Final local verification also passes 107 fast Boost tests / 365 assertions, all 15 MariaDB-backed cases / 43 assertions, PHP-CS-Fixer, syntax, sink inventory, architectural hotspot checks, and staged-source guard. Full CI and a fresh automated re-review are running on this exact head.

**Comment by somethingwithproof** [comment]
The two suppressed operational findings are fixed in commit `17079549b`. Docker-first contract tests failed before editing, proving that recovery status 2 was conditional on success and that rejected archive-table diagnostics omitted the name. Recovery now always returns the remote poller to online status after clearing its PID, while retaining the failure log and nonzero exit code. Invalid archive-like table warnings now include the actual name after control-character sanitization. Post-fix results: 25 focused tests / 126 assertions, 107 fast Boost tests / 367 assertions, full Docker PHPStan across 436 files, PHP-CS-Fixer, syntax, sink inventory, architectural hotspot, and staged-source guards all pass. Full CI and a fresh automated re-review are running on the exact head.

**Comment by somethingwithproof** [comment]
The test-bootstrap finding is fixed in commit `1c76af2f5`. A Docker-first regression confirmed the file-scope `cacti_sizeof()` replacement was still present; it is now removed, the test relies on `tests/bootstrap-unit.php`, and a new contract prevents the stub from returning. Final local results: 15 focused barrier tests / 73 assertions, 108 fast Boost tests / 368 assertions, full Docker PHPStan across 436 files, sink inventory, architectural hotspot checks, and staged-source guard all pass. Full CI and one more exact-head automated review are running.

File: lib/boost.php [hunk 15]
```diff
@@ -469,6 +563,132 @@ function boost_return_cached_image(&$graph_data_array) : bool {
...
+function boost_graph_cache_filename(string $cache_directory, int $local_graph_id, mixed $rra_id, int $timespan, string $business_hours_index, array $graph_data_array) : string {
+	static $secret = null;
+
+	if ($secret === null) {
+		$secret = read_config_option('boost_png_cache_secret');
+
+		if (!is_string($secret) || !preg_match('/^[a-f0-9]{64}$/D', $secret)) {
+			$candidate = bin2hex(random_bytes(32));
+			db_execute_prepared('INSERT IGNORE INTO settings (name, value) VALUES (?, ?)', ['boost_png_cache_secret', $candidate]);
+			$secret = read_config_option('boost_png_cache_secret');
```
[the remark this answers is not in the document]
Comment (somethingwithproof) [reply lib/boost.php]: Fixed in 3c0091564. The post-`INSERT IGNORE` read now bypasses the option cache, so concurrent creators use the database-persisted secret. A hand-off contract assertion covers the forced refresh, and the Docker Boost suite passes 108 tests / 369 assertions.

**Review by TheWitness — APPROVED** [review]

**Review by TheWitness — DISMISSED** [review]

**Review by TheWitness — APPROVED** [review]

**Review by bmfmancini — APPROVED** [review]
